kube-runtime/src/coordination.rs

```diff
@@ -0,0 +1,676 @@
+            if Some(&lease) == self.state.get_lease() {
+                return; // No change in lease, so simply return.
+            }
+            lease
+        } else {
+            self.state = State::Standby;
+            let _res = self.state_tx.send_if_modified(|val| {
+                if matches!(val, LeaderState::Standby) {
+                    false
+                } else {
+                    *val = LeaderState::Standby;
+                    true
+                }
+            });
+            return;
+        };
+
+        // Process the given lease and update as needed.
+        let holder = updated_lease
+            .spec
+            .as_ref()
+            .and_then(|spec| spec.holder_identity.as_deref())
+            .unwrap_or_default();
+        let (is_lease_held, now) = (holder == self.config.identity, Instant::now());
+        match &mut self.state {
+            // In all cases where the lease is held by this identity, we simply update observed info.
+            val @ (State::Leading { .. } | State::Following { .. } | State::Standby) if is_lease_held => {
+                *val = State::Leading {
+                    lease: updated_lease,
+                    last_updated: now,
+                };
+            }
+            // In any case where the holder ID is an empty string, we set to standby, as such a
+            // config is invalid, and indicates that the lease is open for acquisition.
+            val if holder.is_empty() => {
+                *val = State::Standby;
+            }
+            // In all other cases, we are following, and we simply update observed info.
+            val => {
+                *val = State::Following {
+                    leader: holder.into(),
+                    lease: updated_lease,
+                    last_updated: now,
+                };
+            }
+        };
+        self.state_tx.send_if_modified(|val| match &self.state {
+            State::Leading { .. } if !matches!(val, LeaderState::Leading) => {
+                *val = LeaderState::Leading;
+                true
+            }
+            State::Following { .. } if !matches!(val, LeaderState::Following) => {
+                *val = LeaderState::Following;
+                true
+            }
+            State::Standby if !matches!(val, LeaderState::Standby) => {
+                *val = LeaderState::Standby;
+                true
+            }
+            _ => false,
+        });
+        tracing::debug!("lease updated");
+    }
+
+    /// Get the duration to delay before attempting the next lease update.
+    fn get_next_acquire_renew_time(&mut self) -> Duration {
+        // Unpack the last observed change and the configured delay times.
+        let (last_updated, delay) = match &self.state {
+            // If running as leader, then we renew a bit earlier then the lease duration.
+            State::Leading { last_updated, .. } => (last_updated, self.config.renew_deadline),
```

Comment:
@thedodd is this `retry_period` according to the upstream `leaderelection.go`? Otherwise one network failure might let the leader lose its lease. Because after renew_deadline, it is not allowed for the leader to renew the lease according to the protocol.

Reply:
> it is not allowed for the leader to renew the lease according to the protocol.

Do you mind providing a reference for that? I don't recall coming across that constraint, but I'm happy to review and see if there is something which I've missed `:)`.

Reply:
@thedodd Thanks for the reply, and I would be more than happy to have a discussion about this leader election protocol!

I checked the recommended config values upstream, and here they are: lease_duration is 15s, renew_deadline is 10s, retry_period is 2s. So let's take these values as an example.

First of all, if a lease is not renewed within 15s, then the followers are able to compete leadership. While if a current leader cannot renew the lease within 10s, it should't retry(as I quoted from upstream `RenewDeadline is the duration that the acting master will retry refreshing leadership before giving up`). IIUC, the time difference between 10s and 15s is to let the previous leader(or the partitioned leader) to step down leadership and finish some cleanup work as a leader. So when it comes to the time of `15s`, it probably has no leader in the cluster(or at least the possibility is minimized), which means the possibility of two concurrent leaders is minimized.

As of one member case, it can compete leadership after `15s`, so basically not a problem.

I hope we've reached an agreement that leader cannot renew the lease after `renew_deadline`. Then as I tested this PR, the current leader tries to renew the lease at `9s`, which means if this renew request failed, it probably won't get another chance to extend its lease. I think that's why the recommended `renew_deadline` is 5 times than `retry_period`, which is able to tolerate several "heartbeat" failure.

Reply:
> I checked the recommended config values upstream, and here they are: lease_duration is 15s, renew_deadline is 10s, retry_period is 2s.

Agreed. The default configuration here uses the same values (I just wanted to mention that in case you have not seen that yet) `:)`.

> First of all, if a lease is not renewed within 15s, then the followers are able to compete leadership. While if a current leader cannot renew the lease within 10s, it should't retry(as I quoted from upstream `RenewDeadline is the duration that the acting master will retry refreshing leadership before giving up`). IIUC, the time difference between 10s and 15s is to let the previous leader(or the partitioned leader) to step down leadership and finish some cleanup work as a leader.

Ok, I see what you are referring to now. Specifically on the following note:

> IIUC, the time difference between 10s and 15s is to let the previous leader(or the partitioned leader) to step down leadership and finish some cleanup work as a leader

I would argue that this is incorrect, is not the intention of the default 5s difference in config, and that such an assumption could actually lead to quite a bit of trouble. The following are my supporting arguments:
- The K8s documentation is clear that the coordination API (of which this is a part) does not provide a solution for fencing, and as such can make no guarantee that there is only 1 leader at any point in time.<br/>In order to use the coordination API effectively, folks definitely need to let go of the idea that this is something like Raft or Paxos, and it is definitely not something like a mutex which has real capability to enforce ordering of operations or anything like that.
- Instead, the coordination API is a mechanism purely to facilitate "coordination". Multiple different processes working together in a best-effort fashion. 
- Perhaps the most important argument here is this: **there is no protocol**. In reality, the only protocol which K8s provides to make any of this work is the [CC&C guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#concurrency-control-and-consistency). That's it. There is nothing else, no other protocol.
  - Anything folks might want to do with this API, it is important to accept that K8s itself is only providing a nice-to-have to help reduce contention.
  - At the end of the day, it is the data in the backing etcd cluster which actually matters, and the only API to interact with that data is the K8s API (unless you go to etcd directly, which is **_definitely_** a bad idea).
  - It is up to the end-user to code correctly via the CC&C guidelines.

A nice guiding principle from the very first line of the CC&C guideline:

> Kubernetes leverages the concept of resource versions to achieve optimistic concurrency.

At the end of the day, that is what everything boils down to. Leverage resource versions to guard against stale operations. Coordination API just helps reduce superfluous concurrency. I know I said a lot here, but hopefully that is helpful information.

Reply:
And actually, in a more concrete sense, the 5s difference is also just to help ensure that the follower processes do not attempt to acquire too aggressively. Otherwise one would end up with unstable leadership.

Reply:
Did you mean that since there is no protocol, this implementation does not follow exactly the same logic as upstream?

> And as source code indicates, the context timeout is RenewDeadline and interval is RetryPeriod.
>
> https://github.com/kubernetes/client-go/blob/2a6c116e406126324eee341e874612a5093bdbb0/tools/leaderelection/leaderelection.go#L265-L273

If it is so, it's totally fine:) Just need a clarification. Because I believe lots of people like me are expecting a rust implementation with exactly the same logic as it is implemented in Go.

Reply:
> Did you mean that since there is no protocol, this implementation does not follow exactly the same logic as upstream?

Sorry, no. That's not what I mean. Here is my point: holding a lease doesn't actually accomplish anything at all. Any other random system could come in and blow away your lease, it could update any and all data in K8s, the lease doesn't actually give you anything on its own. **Leases are the building block of a protocol, not the protocol itself.**

> If it is so, it's totally fine:) Just need a clarification. Because I believe people like me are expecting a rust implementation with exactly the same logic as it is implemented in Go.

This is the same as the Go implementation, specifically in terms of end result. Obviously, this is a translation from Go to Rust, and not every single detail is the same (or even applies in the same way across languages). However, the end result is the same.

One notable exception here: we do not clear the lease when we lose leadership or fail for some reason. If used as an optimization, we could attempt to release when the leader is shutting down, and followers would be able to see the released lease and immediately attempt to acquire (+ some jitter). I think that is an optimization I would be ok with adding in. However, I would caveat that this is not some reliable abstraction which should be leveraged for doing cleanup work, or something like that.

Reply:
@shuoranliu you've probably seen this, but @clux brought up some discussion on this last note [over here](https://github.com/kube-rs/kube/pull/1056#discussion_r1116521340). Feel free to chime in there if you have some thoughts you would like to share on this front.

Reply:
> Leases are the building block of a protocol, not the protocol itself.

This is true. So that's why I tried to highlight the behaviors of leader and followers in regards to `lease_duration`, `renew_deadline` and `retry_period` in the Go implementation. That the leader renew the lease every `retry_period` and renew is not allowed after `renew_deadline`, which is different from the implementation in this PR. Besides, the name of "renew_deadline" is kind of self-explained, right? The renew is not performed after deadline.

> This is the same as the Go implementation, specifically in terms of end result. Obviously, this is a translation from Go to Rust, and not every single detail is the same (or even applies in the same way across languages). However, the end result is the same.

I'm afraid I cannot agree. The behaviors I stated above in the Go implementation makes a notable difference.

Reply:
> I'm afraid I cannot agree. The behaviors I stated above in the Go implementation makes a notable difference, which is also an important building block of the protocol in my point of view.

Understood. I see you point, and I can agree to an extent. The thing I am wondering though is how much of that is just implementation detail vs actually being a part of the lease acquisition protocol. Deadline is also interpretable from the standpoint of the followers, it doesn't necessitate a final point of retries allowed by the current holder.

For example, if the leader attempts to renew at the 9s mark where the deadline is 10s, but runs into a very small blip from which it could immediately recover, well ... where is the value in refusing to try to renew the lease again just because we've passed the 10s mark? It would ensure that leadership stays stable for a longer period of time, and is overall a more simple path forward. The followers wouldn't try to contend for leadership for another 5s according to the default config anyway.

If we find some consensus here, then I believe that it would leave only the clearing of the lease on step-down as an outstanding item of "real" difference. Thoughts?